custom_components/iotawatt/sensor.py

```diff
@@ -94,13 +107,26 @@ def device_state_attributes(self):
         else:
             channel = "N/A"
 
-        return {"Type": self._io_type, "Channel": channel}
+        attrs = {
+                "type": self._io_type,
+                "channel": channel
+            }
+
+        return attrs
 
     @property
     def state(self):
         """Return the state of the sensor."""
         return self.coordinator.data["sensors"][self._ent].getValue()
 
+    @property
+    def last_reset(self):
+        """Return the time when the sensor was last reset, if any."""
+        last_reset = self.coordinator.data["sensors"][self._ent].getBegin()
+        if last_reset is None:
+            return None
+        return dt.parse_datetime(last_reset)
```

Comment:
@agners The ```last_reset``` attribute in HA 2021.8 is being set to ```2021-01-01T00:00:00```. I guess this is due to the query being sent to the IoTaWatt device asking for WattHours since the beginning of the year.  Though I think this means from the point of view of HA that the sensor will be reset Jan 1 each year.

Looking at the[ IoTaWatt Query API](https://docs.iotawatt.com/en/master/query.html) docs it appears the attribute ```begin``` and ```end``` are required in the URL string. I'm looking to see if there is another way to query IoTaWatt and get  all of the sensor data, but I don't see it. The query is either relative time or specific time.

Beginning relative though is limited to the current year. If I use a specific time such as replacing the ```y``` with ```0000000001``` to use Unix time (1 second from 1970) the DateTime returned by the IoTaWatt device says 1969.

I'll keep looking unless this is an acceptable limitation that HA devs will accept where the last_reset date and value is automatically reset at the beginning of each year?

Reply:
Home Assistant can handle any reset time since we track hourly growth. So when we see a sensor reset, we just note that and adjust the growth accordingly when processing states.

Reply:
@gtdiehl the Query API allows any start date basically. In theory we could use energy usage with a reset date of the last minute. But if we query the API every 30s, there will be a chance that the second query gets in just after the minute mark, leading to ~29s of "lost" energy. This is problematic since we accumulate the differences in HA.

That makes me think that a start date as far back as possible is the best solution.

Let me back up with a bit of history here: I use IoTaWatt since 2019, and was using the REST integration to query it. There I always used the Query API with a begin date of 2019-01-01. However, I recently discovered that IoTaWatt didn't not summarize the values correctly in this case. I discussed it with the creator of IoTaWatt:
https://community.iotawatt.com/t/total-energy-consumption-fluctuation/3149

Gist: Ideally we would have a way to Query just from the beginning of the history log. That is currently not possible. Using `begin=y` (start of the year) is the largest timespan we can define relative and minimizes any accumulation errors.

Reply:
Is there an API endpoint to query the first moment there is data?

Reply:
Not right now. But developers thing about adding this feature by just using the first record as start date when a begin date is before the first record (last post in the above link). We then could just query the API with a far back date. The API will return the date of the first record which can be used as true `last_reset`.